### PR TITLE
Add User itemtype to Licenses linked items

### PR DESCRIPTION
**Pull request by RomainLvr**
## Checklist before requesting a review

*Please delete options that are not relevant.*

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] This change requires a documentation update.

## Description

- It fixes !34567
- Here is a brief description of what this PR does

Adding the relationship of Licenses and Contracts to Users

Linked with [#18388](https://github.com/glpi-project/glpi/pull/18388/files)

## Screenshots (if appropriate):

Licenses : 

image

Contracts : 
image

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
The discussion in #13294 seems to to indicate that this feature was refused last time.

I've also found this on our gitlab: https://gitlab.teclib.com/editions/projects/features-request/-/issues/133

**Comment by cconard96** [comment]
AFAIK this still hasn't been planned out and there is so much related to license management that people have been asking for for years that just isn't possible yet. It only really makes sense to me to do a complete rework.

**Comment by orthagh** [comment]
I don't remember at all the reason why #13294 was refused.
After checking, the important point is the summary tab and the countdown of used licenses and this seems to work correctly.

Maybe there are issues (I didn't check) with linking entities of users, which are managed differently than in other objects.
Also, if we start adding more objects to licenses, we need a dedicated CFG_GLPI constant.

As Curtis says, we still should rework licenses and at least **decouple them from software itemtype**, but regarding assignments to users, I don't see UX issues. As said above, there may exist technical problems.

**Review by stonebuzz — APPROVED** [review]

File: front/item_softwarelicense.form.php [hunk 1]
```diff
@@ -37,7 +37,13 @@
 use Glpi\Exception\Http\BadRequestHttpException;
 
 Session::checkRight("software", UPDATE);
-$isl = new Item_SoftwareLicense();
+
+if (isset($_POST['itemtype']) && $_POST['itemtype'] == 'User') {
+    $isl = new SoftwareLicense_User();
+    $_POST['users_id'] = $_POST['items_id'];
```
Comment (cedric-anne) [inline front/item_softwarelicense.form.php:43]:
```suggestion
    $_POST['users_id'] = $_POST['items_id'];
    unset($_POST['itemtype'], $_POST['items_id']);
```

File: install/migrations/update_10.0.x_to_11.0.0/software_license.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+/**
+ * @var \Migration $migration
+ * @var \DBmysql $DB
+ */
+
+$default_key_sign = DBConnection::getDefaultPrimaryKeySignOption();
+
+if ($DB->fieldExists(SoftwareLicense::getTable(), 'softwares_id')) {
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_11.0.0/software_license.php:43]: Migrations should always use hardcoded names. Same remark for other occurrences in this file.

```suggestion
if ($DB->fieldExists('glpi_softwarelicenses', 'softwares_id')) {
```

File: install/migrations/update_10.0.x_to_11.0.0/software_license.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * ---------------------------------------------------------------------
+ *
+ * GLPI - Gestionnaire Libre de Parc Informatique
+ *
+ * http://glpi-project.org
+ *
+ * @copyright 2015-2024 Teclib' and contributors.
+ * @copyright 2003-2014 by the INDEPNET Development Team.
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_11.0.0/software_license.php:11]: No need to have this copyright line on new files.

```suggestion
```

File: install/mysql/glpi-empty.sql [hunk 4]
```diff
@@ -10074,3 +10074,13 @@ CREATE TABLE `glpi_assets_customfielddefinitions` (
...
+
+DROP TABLE IF EXISTS `glpi_softwarelicenses_users`;
+CREATE TABLE `glpi_softwarelicenses_users` (
+  `id` int unsigned NOT NULL AUTO_INCREMENT,
+  `softwarelicenses_id` int unsigned NOT NULL DEFAULT '0',
+  `users_id` int unsigned NOT NULL DEFAULT '0',
+  PRIMARY KEY (`id`),
+  KEY `item` (`users_id`),
+  KEY `softwarelicenses_id` (`softwarelicenses_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci ROW_FORMAT=DYNAMIC;
```
Comment (cedric-anne) [inline install/mysql/glpi-empty.sql:10086]: Please move this before `SET FOREIGN_KEY_CHECKS=1;`.

File: src/Item_SoftwareLicense.php [hunk 5]
```diff
@@ -752,16 +766,81 @@ function updateItemDropdown(itemtype_el) {
             }
             $queries[] = $query;
         }
+
+        // Add LicensSoftwareLicense_Usere_User
```
Comment (cedric-anne) [inline src/Item_SoftwareLicense.php:770]:
```suggestion
        // Add SoftwareLicense_User
```

File: src/Item_SoftwareLicense.php [hunk 6]
```diff
@@ -752,16 +766,81 @@ function updateItemDropdown(itemtype_el) {
...
+        $entity_table = Entity::getTable();
+
+        $user_query = [
+            'SELECT' => [
+                "$license_users_table.id",
+                "$license_users_table.users_id AS items_id",
+                new QueryExpression($DB::quoteValue(User::class), 'itemtype'),
+                "$license_users_table.softwarelicenses_id",
+                new QueryExpression($DB::quoteValue(0), 'id_deleted'),
+                new QueryExpression($DB::quoteValue(1), 'is_dynamic'),
```
Comment (cedric-anne) [inline src/Item_SoftwareLicense.php:782]:
```suggestion
                new QueryExpression($DB::quoteValue(0), 'is_dynamic'),
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
You should add a `SoftwareLicense_User` tab in the `User` form.

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**